File: Programmers/level2/행렬 테두리 회전하기.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>
using namespace std;
vector<int> answer;
void rotate(int x1, int y1, int x2, int y2, vector<vector<int> >& m, int r, int c) {
    vector<int> nums;
    int tmp = m[x1 - 1][y1 - 1];
    nums.push_back(tmp);
    for (int i = x1 - 1; i < x2 - 1; i++) {
        nums.push_back(m[i + 1][y1 - 1]);
        m[i][y1 - 1] = m[i + 1][y1 - 1];
    }
    for (int i = y1 - 1; i < y2 - 1; i++) {
        nums.push_back(m[x2 - 1][i + 1]);
        m[x2 - 1][i] = m[x2 - 1][i + 1];
    }
    for (int i = x2 - 1; i > x1 - 1; i--) {
        nums.push_back(m[i - 1][y2 - 1]);
        m[i][y2 - 1] = m[i - 1][y2 - 1];
    }
    for (int i = y2 - 1; i > y1; i--) {
        nums.push_back(m[x1 - 1][i - 1]);
        m[x1 - 1][i] = m[x1 - 1][i - 1];
    }
    m[x1 - 1][y1] = tmp;
    answer.push_back(*min_element(nums.begin(), nums.end()));
    return;
}
vector<int> solution(int rows, int columns, vector<vector<int>> queries) {
    int n = 1;
    vector<vector<int> > map(rows, vector<int>(columns, 0));
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < columns; j++) {
            map[i][j] = n++;
        }
    }
    for (int i = 0; i < queries.size(); i++) {
        rotate(queries[i][0], queries[i][1], queries[i][2], queries[i][3], map, rows, columns);
    }
    return answer;
}

File: Programmers/level2/행렬 테두리 회전하기.py
def solution(rows, columns, queries):
    answer = []
    arr=[ [0 for _ in range(columns)] for _ in range(rows)]
    num=1
    for i in range(rows):
        for j in range(columns):
            arr[i][j]=num
            num+=1
            
    for q in queries:
        x1,y1,x2,y2=q
        r,c=x1-1,y1-1
        tmp=arr[r][c]
        min_num=tmp
        
        for mr in range(r+1,x2):
            min_num=min(min_num,arr[mr][c])
            arr[mr-1][c]=arr[mr][c]
        
        
        for mc in range(c+1,y2):
            min_num=min(min_num,arr[x2-1][mc])
            arr[x2-1][mc-1]=arr[x2-1][mc]
        
        for mr in range(x2-2,r-1,-1):
            min_num=min(min_num,arr[mr][y2-1])
            arr[mr+1][y2-1]=arr[mr][y2-1]
        
        for mc in range(y2-2,c-1,-1):
            if mc==c:
                arr[r][mc+1]=tmp
                continue
            min_num=min(min_num,arr[r][mc])
            arr[r][mc+1]=arr[r][mc]
        answer.append(min_num)
        
    
    return answer
